hphp/util/mutex.h

```diff
@@ -326,6 +328,104 @@ struct RankedCHM : tbb::concurrent_hash_map<K, V, H> {
   bool erase(const K& k)                     { return RawCHM::erase(k); }
 };
 
+/*
+ * Need to subclass tbb::chm rather than RankedCHM so we
+ * have access to the protected my_hash_compare member
+ * for getting hashcodes to inject into const SmallLock locks[N].
+ * This is necessary for the erase case where we only have an accessor
+ * with a hashcode_t already inside, and need to be consistent
+ * with how we may have produced a hashcode_t earlier for the same key.
+ */
+#define DEFAULT_CHM_LOCK_BUCKET_COUNT 137
+template<typename K, typename V, typename H=K, Rank R=RankUnranked, uint32_t N=DEFAULT_CHM_LOCK_BUCKET_COUNT>
+struct RankedBucketCHM : tbb::concurrent_hash_map<K, V, H> {
+private:
+  typedef tbb::concurrent_hash_map<K, V, H> RawCHM;
+
+  /*
+   * InternalSmallLock is a union to avoid issues with default initialization of an
+   * array or container of SmallLocks - base_int is never actually used, only lock
+   */
+  typedef union { uint32_t base_int; mutable SmallLock lock; } InternalSmallLock;
+
+  const uint32_t nLocks = N;
+  mutable InternalSmallLock locks[N];
+
+  const size_t outerLockIndex(const K& k) const {
+    return outerLockIndex(RawCHM::my_hash_compare.hash(k));
+  }
+
+  const size_t outerLockIndex(const size_t h) const {
+    return (h % nLocks);
+  }
+public:
+  /*
+   * Same accessors as RankedCHM
+   */
+  struct accessor : RawCHM::accessor {
+    accessor() : freed(false) { pushRank(R); }
+    ~accessor() { if (!freed) popRank(R); }
+    void release() {
+      RawCHM::accessor::release();
+      popRank(R);
+      freed = true;
+    }
+  private:
+    bool freed;
+  };
+  struct const_accessor : RawCHM::const_accessor {
+    const_accessor() : freed(false) { pushRank(R); }
+    ~const_accessor() { if (!freed) popRank(R); }
+    void release() {
+      RawCHM::const_accessor::release();
+      popRank(R);
+      freed = true;
+    }
+  private:
+    bool freed;
+  };
+
+  /*
+   * All find/insert/erase operations hold an outer lock, which prevents
+   * many threads from spinning aggressively on the spin_rw_mutex inside
+   * of tbb::concurrent_hash_map
+   */
+  bool find(const_accessor& a, const K& k) const {
+    std::unique_lock<SmallLock> outerLock(locks[outerLockIndex(k)].lock);
```

Comment:
I'm no longer sure I understand why this works :(.

I was imagining the the outer critical section would coincide with the lifetime of the accessor. With this scheme, it seems like it should be possible (?) for a caller to `find(a, k)`, get an `a` that points to nothing and is a lock, and then have any number of readers also `find(a, k)` and burn CPU in the body of `tbb::chm::find()` waiting for the writer to destroy its accessor.

What am I missing?

Reply:
OK. While I see why we're seeing the benefit now, I feel like it would be easier to reason about, more robust, and possibly free up that one extra core at T2 in Jamie's writeup, to scope the SmallLock critical section to the accessor lifetime (or, accessor acquire/release() span anyway). Does that seem reasonable?